provider/openstack/firewaller.go

```diff
@@ -545,92 +545,126 @@ func (c *neutronFirewaller) setUpGlobalGroup(groupName string, apiPort int) (neu
+	var group neutron.SecurityGroupV2
+
 	// First attempt to look up an existing group by name.
 	groupsFound, err := neutronClient.SecurityGroupByNameV2(name)
-	if err == nil {
-		for _, group := range groupsFound {
-			if c.verifyGroupRules(group.Rules, rules) {
-				return group, nil
-			}
+	// a list is returned, but there should be only one
+	if err == nil && len(groupsFound) == 1 {
+		group = groupsFound[0]
+	} else if strings.Contains(err.Error(), "failed to find security group") {
+		g, err := neutronClient.CreateSecurityGroupV2(name, "juju group")
+		if err != nil {
+			return zeroGroup, err
 		}
-	}
-	// Doesn't exist, so try and create it.
-	newGroup, err := neutronClient.CreateSecurityGroupV2(name, "juju group")
-	if err != nil {
+		group = *g
+	} else {
 		return zeroGroup, err
 	}
-	// The new group is created so now add the rules.
-	for _, rule := range rules {
-		rule.ParentGroupId = newGroup.Id
+
+	have := newRuleInfoSetFromRules(group.Rules)
+	want := newRuleInfoSetFromRuleInfo(rules)
+
+	// Find rules we want to delete, that we have but don't want, and
+	// delete them.
+	remove := make(ruleInfoSet)
+	for k := range have {
+		// Neutron creates 2 egress rules with any new Security Group.
+		// Keep them.
+		if _, ok := want[k]; !ok && k.Direction != "egress" {
+			remove[k] = have[k]
+		}
+	}
+	for _, ruleId := range remove {
+		if err = neutronClient.DeleteSecurityGroupRuleV2(ruleId); err != nil {
+			return zeroGroup, err
+		}
+	}
+
+	// Find rules we want to add, that we want but don't have, and add
+	// them.
+	add := make(ruleInfoSet)
+	for k := range want {
+		if _, ok := have[k]; !ok {
+			add[k] = want[k]
+		}
+	}
+	for rule, _ := range add {
+		rule.ParentGroupId = group.Id
 		// Neutron translates empty RemoteIPPrefix into
 		// 0.0.0.0/0 or ::/0 instead of ParentGroupId
 		// when EthernetType is set
 		if rule.RemoteIPPrefix == "" {
-			rule.RemoteGroupId = newGroup.Id
+			rule.RemoteGroupId = group.Id
 		}
-		groupRule, err := neutronClient.CreateSecurityGroupRuleV2(rule)
-		if err != nil {
+		if _, err := neutronClient.CreateSecurityGroupRuleV2(rule); err != nil {
 			return zeroGroup, err
 		}
-		newGroup.Rules = append(newGroup.Rules, *groupRule)
 	}
-	return *newGroup, nil
-}
 
-func countIngressRules(rules []neutron.SecurityGroupRuleV2) int {
-	count := 0
-	for _, rule := range rules {
-		if rule.Direction == "ingress" {
-			count += 1
-		}
-	}
-	return count
-}
-
-// verifyGroupRules verifies the group rules against the rules we're looking for.
-func (c *neutronFirewaller) verifyGroupRules(rules []neutron.SecurityGroupRuleV2, rulesToMatch []neutron.RuleInfoV2) bool {
-	if countIngressRules(rules) != len(rulesToMatch) {
-		return false
+	// Since we may have done a few add or delete rules, get a new
+	// copy of the security group to return containing the end
+	// list of rules.
+	groupsFound, err = neutronClient.SecurityGroupByNameV2(name)
+	if err != nil || !(err == nil && len(groupsFound) == 1) {
```

Comment:
this can better be expressed as
if err != nil || len(groupsFound) > 1 

however

I'd like us to return an explicit error for len(groupsFound) > 1
The code as is will return zeroGroup, nil, which is bad. The caller wil think that there's no error but not get a usable group. That breaks the contract.